clients/command/src/main/java/org/eclipse/hono/client/command/KubernetesContainerInfoProvider.java

```diff
@@ -0,0 +1,230 @@
+     * is running in contains multiple running containers.
+     */
+    public static final String KUBERNETES_CONTAINER_NAME_ENV_VAR = "KUBERNETES_CONTAINER_NAME";
+
+    private static final Logger LOG = LoggerFactory.getLogger(KubernetesContainerInfoProvider.class);
+
+    private static final KubernetesContainerInfoProvider INSTANCE = new KubernetesContainerInfoProvider();
+
+    private final AtomicReference<Promise<String>> containerIdPromiseRef = new AtomicReference<>();
+
+    private String containerId;
+    private final String podName;
+    private final boolean isRunningInKubernetes;
+
+    private KubernetesContainerInfoProvider() {
+        isRunningInKubernetes = System.getenv("KUBERNETES_SERVICE_HOST") != null;
+        podName = isRunningInKubernetes ? System.getenv("HOSTNAME") : null;
+    }
+
+    /**
+     * Gets the KubernetesContainerInfoProvider instance.
+     *
+     * @return The KubernetesContainerInfoProvider.
+     */
+    public static KubernetesContainerInfoProvider getInstance() {
+        return INSTANCE;
+    }
+
+    boolean isRunningInKubernetes() {
+        return isRunningInKubernetes;
+    }
+
+    /**
+     * Gets the Kubernetes pod name if running in Kubernetes.
+     *
+     * @return The pod name or {@code null} if not running in Kubernetes.
+     */
+    public String getPodName() {
+        return podName;
+    }
+
+    /**
+     * Determines the container id if running in a container in Kubernetes.
+     * <p>
+     * First an attempt is made to get the container id by inspecting <code>/proc/self/cgroup</code>
+     * (via {@link CgroupV1KubernetesContainerUtil#getContainerId()}).
+     * If not found there, the container id is queried via the Kubernetes API.
+     * <p>
+     * NOTE: The service account of the application pod must have an RBAC role allowing "get" on the "pods" resource.
+     * If this application is running in a pod with multiple containers, the container that this application is running
+     * in must have an environment variable with the name specified in {@link #KUBERNETES_CONTAINER_NAME_ENV_VAR} set
+     * to the container name.
+     *
+     * @return A future indicating the outcome of the operation.
+     *         The future will be succeeded with the container id or {@code null} if not running in Kubernetes.
+     *         In case of an error determining the container id, the future will be failed with either a
+     *         {@link IllegalStateException} if a precondition for getting the id isn't fulfilled (because of missing
+     *         permissions or because multiple pod containers exist and no KUBERNETES_CONTAINER_NAME env var is set),
+     *         or otherwise a {@link ServerErrorException}.
+     */
+    public Future<String> getContainerId() {
+        if (containerId != null) {
+            return Future.succeededFuture(containerId);
+        }
+        final Context currentContext = Vertx.currentContext();
```

Comment:
How about passing a Context into the method instead of relying on the static method to obtain the context? That way, it would also be much easier to test this method, right?

Or maybe it would make even more sense to pass a Vertx instance to the *getInstance* method?

Reply:
I've now added a `context` parameter to the `getContainerId()` method.